src/v2/auth.rs

```diff
@@ -35,91 +173,105 @@ impl Client {
             .await?;
 
         trace!("GET '{}' status: {:?}", r.url(), r.status());
-        let a = r
-            .headers()
+        r.headers()
             .get(reqwest::header::WWW_AUTHENTICATE)
-            .ok_or_else(|| Error::from("get_token: missing Auth header"))?;
-        let chal = String::from_utf8(a.as_bytes().to_vec())?;
+            .ok_or_else(|| {
+                Error::from(format!(
+                    "missing {:?} header",
+                    reqwest::header::WWW_AUTHENTICATE
+                ))
+            })
+            .map(ToOwned::to_owned)
+    }
 
-        let (mut auth_ep, service) = parse_hdr_bearer(chal.trim_start_matches("Bearer "))?;
+    /// Perform registry authentication and return the authenticated client.
+    ///
+    /// If Bearer authentication is used the returned client will be authorized for the requested scopes.
+    pub async fn authenticate(mut self, scopes: &[&str]) -> Result<Self> {
+        let credentials = if let Some(credentials) = self.credentials.clone() {
+            credentials
+        } else {
+            bail!("cannot authenticate without credentials");
+        };
 
-        trace!("Token provider: {}", auth_ep);
-        if let Some(sv) = service {
-            auth_ep += &format!("?service={}", sv);
-            trace!("Service identity: {}", sv);
-        }
+        self.auth = None;
 
-        Ok(auth_ep)
-    }
+        let authentication_header = self.get_www_authentication_header().await?;
+        match WwwAuthenticateHeaderContent::from_www_authentication_header(authentication_header)? {
```

Comment:
This logic is not very nice to follow. I would suggest:
 * doing the `self.auth` assignment in a single place, at the bottom of the function
 * move out the logic in this match statement, from the Bearer arm into its own async method attached to `BearerAuth`